src/libcore/alloc/mod.rs

```diff
@@ -0,0 +1,337 @@
+//! Memory allocation APIs
+
+#![stable(feature = "alloc_module", since = "1.28.0")]
+
+mod global;
+mod layout;
+
+#[stable(feature = "global_alloc", since = "1.28.0")]
+pub use self::global::GlobalAlloc;
+#[stable(feature = "alloc_layout", since = "1.28.0")]
+pub use self::layout::{Layout, LayoutErr};
+
+use crate::cmp;
+use crate::fmt;
+use crate::intrinsics;
+use crate::ptr::{self, NonNull};
+
+/// The `AllocErr` error indicates an allocation failure
+/// that may be due to resource exhaustion or to
+/// something wrong when combining the given input arguments with this
+/// allocator.
+#[unstable(feature = "allocator_api", issue = "32838")]
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct AllocErr;
+
+// (we need this for downstream impl of trait Error)
+#[unstable(feature = "allocator_api", issue = "32838")]
+impl fmt::Display for AllocErr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("memory allocation failed")
+    }
+}
+
+/// A desired initial state for allocated memory.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[unstable(feature = "allocator_api", issue = "32838")]
+pub enum AllocInit {
+    /// The new memory is not initialized to any particular value.
+    ///
+    /// Remember that reading uninitialized memory is Undefined Behavior.
+    Uninitialized,
+    /// The new memory is guaranteed to be zeroed.
+    Zeroed,
+}
+use AllocInit::*;
```

Comment:
Since we no longer provide a default implementation of `alloc_zeroed`, the allocator author now needs to handle zero-initialization themselves. We could ease this by adding a method to `AllocInit` which performs the necessary initialization given a pointer and a layout:

```rust
impl AllocInit {
    pub unsafe fn initialize(self, ptr: NonZero<u8>, layout: Layout) { ... }
}
```